src/write.rs

```diff
@@ -476,7 +483,177 @@ where
+                                crate::TryCompressResult::Independent(re_next) => {
+                                    next = re_next;
+                                    // We consumed one of our range and need to check whether to break or continue.
+                                    range_remaining -= 1;
+                                    if range_remaining == 0 {
+                                        break;
+                                    } else {
+                                        continue;
+                                    }
+                                }
+                                // prev must precede next: restore next then break
+                                crate::TryCompressResult::Dependent(re_next) => {
+                                    next = re_next;
+                                    break;
+                                }
+                            }
+                        } else {
+                            // Remember empty loc for efficient insertion
+                            none.replace((prev_rev_idx, prev_loc));
+                            // If the empty loc is at the back of the non-none oplog we can increment rev_dirty_range.start.
+                            if prev_rev_idx == rev_dirty_range.start {
+                                rev_dirty_range.start += 1;
+                            }
+                        }
+                    }
+                    // found nothing to combine with / encountered dependency
+                    // See if we found an empty loc during iteration, else push
+                    if let Some((none_rev_idx, none_loc)) = none {
+                        if none_loc.replace(next).is_some() {
+                            debug_assert!(
+                                false,
+                                "Should still have been None since we didn't insert yet."
+                            )
+                        }
+                        // If we inserted before the end of the non-none oplog we need to decrease rev_dirty_range.start.
+                        if none_rev_idx < rev_dirty_range.start {
+                            rev_dirty_range.start = none_rev_idx;
+                        }
+                        // If we inserted at the front of the all-some oplog we can decrement rev_dirty_range.end.
+                        if none_rev_idx + 1 == rev_dirty_range.end {
+                            rev_dirty_range.end = none_rev_idx;
+                        }
+                    } else {
+                        debug_assert!(rev_dirty_range.start == 0, "We only push if we found no nones and we always find at least one none at the back if it is there.");
+                        debug_assert!(rev_dirty_range.end == 0, "Pushing means no nones.");
+                        self.oplog.push_back(Some(next));
+                    }
+                }
+                // stably remove temporary nones from the oplog
```

Comment:
I _think_ you can just `.sort_by_key(|o| o.is_none())` here, and then trim by the number of `None`s. That would save you much of the code below. You could probably also use `rev_dirty_range` to only sort the subslice that you know isn't all `None`s.

Reply:
`.sort_by_key(|o| o.is_none())` would be O(nLogn) instead of `.retain(|o| o.is_some()) with O(n), as well as possibly unstable. And because VecDeque doesn't let us sublice it, both would always be unnecessarily iterating the full oplog, even for only a single None at the back. Though I have effectively reimplemented `retain`, this version only iterates over the section of it that is actually mixed.